File: final_assignment_front/lib/features/ai/business_chat_agent.dart
import 'dart:convert';

import 'package:final_assignment_front/config/routes/app_routes.dart';
import 'package:final_assignment_front/core/auth/role_utils.dart';
import 'package:final_assignment_front/features/model/chat_action.dart';
import 'package:final_assignment_front/features/model/chat_action_response.dart';

class BusinessChatAgent {
  const BusinessChatAgent();

  ChatActionResponse? resolve({
    required String message,
    required String role,
  }) {
    final normalized = message.trim().toLowerCase();
    if (normalized.isEmpty) return null;

    final prefill = _extractPrefill(message);
    if (RoleUtils.isSuperAdminRole(role)) {
      return _resolveSuperAdmin(normalized, prefill) ??
          _resolveAdmin(normalized, prefill);
    }
    if (RoleUtils.isAdminRole(role)) {
      return _resolveAdmin(normalized, prefill);
    }
    return _resolveDriver(normalized, prefill);
  }

  ChatActionResponse? _resolveDriver(
    String message,
    Map<String, String> prefill,
  ) {
    if (_containsAny(message, const ['申诉', '异议', '复议', '撤销处罚'])) {
      return _single(
        answer: '可以进入用户申诉页面，我会把识别到的车牌或违法编号作为预填线索带过去。',
        label: '打开用户申诉',
        route: Routes.userAppeal,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['罚款', '缴费', '缴款', '缴纳', '支付'])) {
      return _single(
        answer: '可以进入罚款缴纳页面，查看待缴记录并继续支付流程。',
        label: '打开罚款缴纳',
        route: Routes.fineInformation,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['违法', '违章', '扣分', '记分', '处罚'])) {
      return _single(
        answer: '可以进入违法详情页面，查看个人违法记录、处理状态和关联车辆信息。',
        label: '查看违法详情',
        route: Routes.userOffenseListPage,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['车辆', '车牌', '登记', '绑定', '行驶证'])) {
      return _single(
        answer: '可以进入车辆登记管理页面，维护车辆档案和绑定信息。',
        label: '打开车辆登记',
        route: Routes.vehicleManagement,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['进度', '消息', '办理状态', '处理结果'])) {
      return _single(
        answer: '可以进入进度消息页面，查看申诉、缴费或业务办理状态。',
        label: '查看进度消息',
        route: Routes.onlineProcessingProgress,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['资料', '个人信息', '身份证', '驾驶证', '手机号'])) {
      return _single(
        answer: '可以进入个人资料页面，维护身份信息、驾驶证信息和联系方式。',
        label: '打开个人资料',
        route: Routes.personalMain,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['地图', '位置', '附近', '导航'])) {
      return _single(
        answer: '可以打开地图模块查看位置相关信息。',
        label: '打开地图',
        route: Routes.map,
        prefill: prefill,
      );
    }
    return null;
  }

  ChatActionResponse? _resolveAdmin(
    String message,
    Map<String, String> prefill,
  ) {
    if (_containsAny(message, const ['申诉', '审批', '审核', '复核'])) {
      return _single(
        answer: '可以进入申诉审批管理页面，处理用户提交的申诉材料和审核进度。',
        label: '打开申诉审批',
        route: Routes.appealManagement,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['扣分', '记分'])) {
      return _single(
        answer: '可以进入扣分管理页面，核对违法扣分记录和处理状态。',
        label: '打开扣分管理',
        route: Routes.deductionManagement,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['罚款', '缴费', '缴款', '缴纳'])) {
      return _single(
        answer: '可以进入罚款管理页面，维护罚款记录、缴款状态和处理人信息。',
        label: '打开罚款管理',
        route: Routes.fineList,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['司机', '驾驶员', '驾驶证'])) {
      return _single(
        answer: '可以进入司机管理页面，查看驾驶员档案及关联车辆、违法、罚款和申诉统计。',
        label: '打开司机管理',
        route: Routes.driverList,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['车辆', '车牌', '车架', '发动机'])) {
      return _single(
        answer: '可以进入车辆管理页面，查看车辆档案、驾驶员绑定和关联业务统计。',
        label: '打开车辆管理',
        route: Routes.vehicleList,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['违法', '违章', '处罚', '违法行为'])) {
      return _single(
        answer: '可以进入违法行为管理页面，查看和维护违法记录。',
        label: '打开违法管理',
        route: Routes.offenseList,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['进度', '消息', '处理状态', '办理状态'])) {
      return _single(
        answer: '可以进入进度管理页面，查看业务请求、幂等状态和处理结果。',
        label: '打开进度管理',
        route: Routes.progressManagement,
        prefill: prefill,
      );
    }
    return null;
  }

  ChatActionResponse? _resolveSuperAdmin(
    String message,
    Map<String, String> prefill,
  ) {
    if (_containsAny(message, const ['rag', '知识库', '知识', '资料录入', '向量'])) {
      return _single(
        answer: '可以进入 RAG 资料管理页面，维护知识资料和向量化任务。',
        label: '打开 RAG 管理',
        route: Routes.ragManagement,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['日志', '审计', '操作记录', '登录记录'])) {
      return _single(
        answer: '可以进入日志管理页面，审查登录日志、操作日志和系统日志。',
        label: '打开日志管理',
        route: Routes.logManagement,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['用户', '账号', '账户', '权限', '角色'])) {
      return _single(
        answer: '可以进入用户管理页面，维护账号状态、角色和权限。',
        label: '打开用户管理',
        route: Routes.userManagementPage,
        prefill: prefill,
      );
    }
    if (_containsAny(message, const ['系统治理', '治理', '异常链路', '运维'])) {
      return _single(
        answer: '可以进入系统治理页面，集中查看系统治理入口和异常链路线索。',
        label: '打开系统治理',
        route: Routes.systemGovernance,
        prefill: prefill,
      );
    }
    return null;
  }

  ChatActionResponse _single({
    required String answer,
    required String label,
    required String route,
    required Map<String, String> prefill,
  }) {
    return ChatActionResponse(
      answer: answer,
      actions: [
        ChatAction(
          type: 'NAVIGATE',
          label: label,
          target: route,
          value: jsonEncode({
            'agentPrefill': prefill,
            'source': 'ai_chat',
          }),
        ),
      ],
      needConfirm: false,
    );
  }

  Map<String, String> _extractPrefill(String message) {
    final prefill = <String, String>{};
    final plate = RegExp(
      r'[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼][A-Z][A-Z0-9]{5,6}',
      caseSensitive: false,
    ).firstMatch(message);
    if (plate != null) {
      prefill['licensePlate'] = plate.group(0)!.toUpperCase();
    }

    final number = RegExp(
      r'(?:违法|违章|处罚|申诉|业务|编号|单号|记录)[：:\s#-]*([A-Za-z0-9]{5,})',
      caseSensitive: false,
    ).firstMatch(message);
    if (number != null) {
      prefill['businessNumber'] = number.group(1)!;
    }
    return prefill;
  }

  bool _containsAny(String message, List<String> keywords) {
    return keywords.any(message.contains);
  }
}

File: final_assignment_front_react/src/api/businessChatAgent.ts
/**
 * 本地业务动作解析器，对齐 Flutter BusinessChatAgent。
 * 在发起 AI 流式请求前，根据关键词命中把消息解析为一条可执行动作
 * （跳转到对应业务页面并携带预填信息）。未命中关键词时返回 null，
 * 由调用方继续走真正的 AI 流式回答。
 *
 * 路由名对齐 Flutter Routes.* 与 React 路由表（见 App.tsx）。
 * 预填字段：licensePlate（车牌）、businessNumber（业务编号）。
 */
import type { ChatAction, ChatActionResponse } from "../api/aiChat";

type Role = "USER" | "ADMIN" | "SUPER_ADMIN";

interface Prefill {
  licensePlate?: string;
  businessNumber?: string;
  source: "ai_chat";
}

const PLATE_RE =
  /[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼][A-Z][A-Z0-9]{5,6}/g;
const BUSINESS_NO_RE =
  /(?:违法|违章|处罚|申诉|业务|编号|单号|记录)[:\s#-]*([A-Za-z0-9]{5,})/g;

function extractPrefill(message: string): Prefill {
  const prefill: Prefill = { source: "ai_chat" };
  const plateMatch = message.toUpperCase().match(PLATE_RE);
  if (plateMatch && plateMatch.length > 0) {
    prefill.licensePlate = plateMatch[0];
  }
  const businessMatch = BUSINESS_NO_RE.exec(message);
  if (businessMatch && businessMatch[1]) {
    prefill.businessNumber = businessMatch[1];
  }
  return prefill;
}

function containsAny(text: string, keywords: string[]): boolean {
  return keywords.some((keyword) => text.includes(keyword));
}

interface ActionSpec {
  label: string;
  target: string;
  keywords: string[];
}

const DRIVER_ACTIONS: ActionSpec[] = [
  {
    label: "打开我的申诉",
    target: "/userAppeal",
    keywords: ["申诉", "异议", "复议", "撤销处罚"],
  },
  {
    label: "打开罚款缴纳",
    target: "/fineInformation",
    keywords: ["罚款", "缴费", "缴款", "缴纳", "支付"],
  },
  {
    label: "打开违法记录",
    target: "/userOffenseListPage",
    keywords: ["违法", "违章", "扣分", "记分", "处罚"],
  },
  {
    label: "打开车辆管理",
    target: "/vehicleManagement",
    keywords: ["车辆", "车牌", "登记", "绑定", "行驶证"],
  },
  {
    label: "打开进度消息",
    target: "/onlineProcessingProgress",
    keywords: ["进度", "消息", "办理状态", "处理结果"],
  },
  {
    label: "打开个人资料",
    target: "/personalMain",
    keywords: ["资料", "个人信息", "身份证", "驾驶证", "手机号"],
  },
  {
    label: "打开服务地图",
    // 对齐 Flutter Routes.map = /admin/map（仅注册在 AdminLayout 下）
    target: "/admin/map",
    keywords: ["地图", "位置", "附近", "导航"],
  },
];

const ADMIN_EXTRA: ActionSpec[] = [
  {
    label: "打开申诉审批",
    target: "/appealManagement",
    keywords: ["申诉", "审批", "审核", "复核"],
  },
  {
    label: "打开扣分管理",
    target: "/deductionManagement",
    keywords: ["扣分", "记分"],
  },
  {
    label: "打开罚款管理",
    target: "/fineList",
    keywords: ["罚款", "缴费", "缴款", "缴纳"],
  },
  {
    label: "打开驾驶员管理",
    target: "/driverList",
    keywords: ["司机", "驾驶员", "驾驶证"],
  },
  {
    label: "打开车辆管理",
    target: "/vehicleList",
    keywords: ["车辆", "车牌", "车架", "发动机"],
  },
  {
    label: "打开违法管理",
    target: "/offenseList",
    keywords: ["违法", "违章", "处罚", "违法行为"],
  },
  {
    label: "打开业务进度",
    target: "/progressManagement",
    keywords: ["进度", "消息", "处理状态", "办理状态"],
  },
];

const SUPER_ADMIN_EXTRA: ActionSpec[] = [
  {
    label: "打开 RAG 资料管理",
    target: "/admin/ragManagement",
    keywords: ["rag", "知识库", "知识", "资料录入", "向量"],
  },
  {
    label: "打开日志审查",
    target: "/admin/logManagement",
    keywords: ["日志", "审计", "操作记录", "登录记录"],
  },
  {
    label: "打开用户与权限",
    target: "/admin/userManagementPage",
    keywords: ["用户", "账号", "账户", "权限", "角色"],
  },
  {
    label: "打开系统治理",
    target: "/admin/systemGovernance",
    keywords: ["系统治理", "治理", "异常链路", "运维"],
  },
];

function resolveFromSpecs(
  message: string,
  specs: ActionSpec[]
): ChatActionResponse | null {
  for (const spec of specs) {
    if (containsAny(message, spec.keywords)) {
      const prefill = extractPrefill(message);
      const action: ChatAction = {
        type: "NAVIGATE",
        label: spec.label,
        target: spec.target,
        value: JSON.stringify({ agentPrefill: prefill, source: "ai_chat" }),
      };
      return {
        answer: "已识别到可执行业务动作，请点击下方按钮继续。",
        actions: [action],
        needConfirm: false,
      };
    }
  }
  return null;
}

/**
 * 解析用户消息为业务动作。对齐 Flutter BusinessChatAgent.resolve：
 * - SUPER_ADMIN 优先尝试超级管理员动作，再退回管理员动作；
 * - ADMIN 使用管理员动作；
 * - 其余使用驾驶员动作。
 * 未命中返回 null。
 */
export function resolveBusinessAction(
  message: string,
  role: Role
): ChatActionResponse | null {
  const normalized = message.trim().toLowerCase();
  if (!normalized) return null;

  if (role === "SUPER_ADMIN") {
    return (
      resolveFromSpecs(normalized, SUPER_ADMIN_EXTRA) ||
      resolveFromSpecs(normalized, ADMIN_EXTRA) ||
      resolveFromSpecs(normalized, DRIVER_ACTIONS)
    );
  }
  if (role === "ADMIN") {
    return resolveFromSpecs(normalized, ADMIN_EXTRA);
  }
  return resolveFromSpecs(normalized, DRIVER_ACTIONS);
}

/** 解析动作 value JSON，提取预填字段，对齐 Flutter _decodeActionArguments。 */
export function decodeActionValue(
  value: string | undefined
): { agentPrefill: Record<string, unknown>; [key: string]: unknown } {
  if (!value) return { agentPrefill: {} };
  try {
    const decoded = JSON.parse(value) as Record<string, unknown>;
    if (decoded && typeof decoded === "object") {
      return {
        agentPrefill:
          (decoded.agentPrefill as Record<string, unknown>) || {},
        ...decoded,
      };
    }
  } catch {
    /* ignore */
  }
  return { agentPrefill: { value } };
}
